Float_u stdlib module

ccasin opened the pull request:
This adds a `Float_u` library.  This is done similarly to the `IArray` library, in that it is compiled as part of stdlib but is not actually in the `Stdlib` module - you access it as `Stdlib__Float_u`.

The diff is useless until the unboxed floats PR is merged.  Instead, look at the last three commits, which:

1. Add the library (which doesn't build).
2. Fix the build (dealing with the problem that `stdlib` is not otherwise compiled with `-extension layouts_alpha`).
3. Add tests.  This takes the approach outlined in ocaml-flambda/ocaml-jst#71, though a bunch of changes and extensions were needed.

The original plan was to give the new stdlib module a name like `Will_miscompile_float_u` or `For_ceastlund's_eyes_only_float_u`, because it's theoretically a way to get at `float#` without using `layouts_alpha`.  But since that plan was developed we've decided to add layout checks to `value_kind` that (I think) will result in rejecting any practical use of this library without the flag.  So I don't think the scary name is needed, but I'm happy to go back to it if preferred.

**Review**: @goldfirere to review or delegate.

goldfirere commented on ocaml/Makefile.common-jst:
```diff
@@ -172,6 +173,12 @@ _install: compiler
 install: _install
 	mkdir -p '$(prefix)'
 	rsync --chmod=u+rw,go+r -rl _install/ '$(prefix)'
+	rm '$(prefix)/lib/ocaml/ocaml_compiler_internal_params'
```
This is all unfortunate (as we've discussed). I've put it on the agenda for this week's group meeting, to see if we can find a better way. And I've made an internal ticket.

goldfirere commented on ocaml/stdlib/float_u.ml:
```diff
@@ -0,0 +1,156 @@
+
+external of_float : (float[@local_opt]) -> float# = "%unbox_float"
+
+(* CR layouts: Investigate whether it's worth making these things externals.
```
We have an internal ticket to come back to this point. Not sure the CR is helping beyond that.

ccasin replied:
I think in a situation like this, where the CR is related to a natural question someone might have reading the code, it's nice to have here.  But I can change it from a CR to a regular comment if you prefer.

goldfirere replied:
No, it's fine. You've identified how it's helping.

goldfirere commented on ocaml/stdlib/float_u.mli:
```diff
@@ -0,0 +1,344 @@
+
+open! Stdlib
+
+(* CR layouts v5: This file is based on [float.mli], which itself is generated
```
v4

goldfirere commented on ocaml/stdlib/float_u.mli:
```diff
@@ -0,0 +1,344 @@
+(** [pred x] returns the floating-point number right before [x] i.e.,
+   the greatest floating-point number smaller than [x].  See also
+   {!next_after}. *)
+
```
No need to fix this in retrospect, but I think it would be a bit better to keep the unused definitions from float.mli here, but commented out. This would make auditing completion easier and would also make the v5 work easier. (Fixing this now would basically be doing that work, so no need.)

goldfirere commented on ocaml/stdlib/float_u.mli:
```diff
@@ -0,0 +1,344 @@
+val max : float# -> float# -> float#
+(** [max x y] returns the maximum of [x] and [y].  It returns [nan]
+    when [x] or [y] is [nan].  Moreover [max (-0.) (+0.) = +0.] *)
+
```
```
(* CR layouts v5: Add [min_max] and [min_max_num] when we have unboxed types in tuples. *)
```

ccasin replied:
(These are already mentioned above)

goldfirere commented on ocaml/stdlib/float_u.mli:
```diff
@@ -0,0 +1,344 @@
+# 1 "float_u.mli"
```
Modules in the stdlib should normally be exposed through the MODULE ALIASES section of stdlib.ml{,i}. You have not done that here. Is that intentional? I could see arguments either way, but actually I think we should expose the module, given our safety check.

You may wish to double-check the instructions in https://github.com/ocaml-flambda/flambda-backend/blob/main/ocaml/stdlib/HACKING.adoc

ccasin replied:
I think the plan was to do what we did for `IArray` (i.e., explicitly not include an alias in the main `stdlib` module).  I'm not actually sure what the reasons to do that in the case of `IArray` are, and agree there isn't a super strong safety reason, but perhaps there were reasons related to public release libraries?

ccasin replied:
@antalsz Can you remind us what the reasoning was for `IArray`?

goldfirere replied:
Hmm...... public release is a good reason. Putting in the `Stdlib__` is enough of a speed bump that people won't do this accidentally, which would be annoying in a public-release library. Let's go with that. Would be good to document all this somewhere.

ccasin replied:
I've added a comment.

goldfirere commented on ocaml/testsuite/tests/typing-layouts-float64/stdlib_float_u_alpha.ml:
```diff
@@ -0,0 +1,269 @@
+(* TEST
```
Very pleasant reading here. I had wanted to create a whiz-bang QuickCheck clone to do this, but you've done something simpler and quicker, and yet still nicely readable. :+1:

goldfirere replied:
Er.... that's a bit embarrassing, really. I had no sense of familiarity. :grimacing: In any case, please trust that I wasn't trying to compliment myself!

goldfirere reviewed (approved):
Just a few small comments.

The tests here test for correctness of the functions. Great. Are there any other properties? Allocations would be nice, but maybe already covered elsewhere (I don't see a compelling need to check each function for allocation separately.) I think that's it, but I wanted to pause to contemplate for a moment.

ccasin commented:
I've integrated review feedback, rebased this onto the latest version of unboxed floats, and added one more commit that changes some tests to use the new stdlib module rather than reimplementing it.

@mshinwell can you approve for the build system hacks?

mshinwell reviewed (approved):
The build system parts are ok
